systemd/jenkins.sh

```diff
@@ -0,0 +1,191 @@
+#!/bin/sh
+
+die() {
+	echo "$(basename "$0"): $*" >&2
+	exit 1
+}
+
+check_env() {
+	required=true
+	for var in "$@"; do
+		if [ "${var}" = '--' ]; then
+			required=false
+			continue
+		fi
+
+		val=$(eval echo "\$${var}")
+		if $required && [ -z "${val}" ]; then
+			die "check_env: ${var} must be non-empty"
+		fi
+	done
+}
+
+infer_java_cmd() {
+	if [ -n "${JENKINS_JAVA_CMD}" ] && [ -x "${JENKINS_JAVA_CMD}" ]; then
+		return 0
+	fi
+
+	if [ -n "${JAVA_HOME}" ] && [ -x "${JAVA_HOME}/bin/java" ]; then
+		JENKINS_JAVA_CMD="${JAVA_HOME}/bin/java"
+		return 0
+	fi
+
+	JENKINS_JAVA_CMD="$(command -v java)" || return "$?"
+}
+
+check_java_version() {
+	java_version=$("${JENKINS_JAVA_CMD}" -version 2>&1 |
+		sed -n ';s/.* version "\([0-9]\{2,\}\|[0-9]\.[0-9]\)\..*".*/\1/p;')
+
+	if [ -z "${java_version}" ]; then
+		return 1
+	elif [ "${java_version}" != "11" ] && [ "${java_version}" != "1.8" ]; then
```

Comment:
This provides no way for the administrator to use the `--enable-future-java` option to allow them to experiment with Java 17.  Is that intentional?  I thought that the previous code allowed non-matching Java versions if enable-future-java was passed in

Reply:
> Is that intentional?

A rhetorical question if I ever saw one. Obviously you see a problem. Why don't you just be direct and say: "That seems wrong to me because … (insert reasoning here)". Note that _guessing_ about what the previous code supported (rather than _confirming_ through testing) is a claim, not an argument.

I haven't tested the old code with Java 17. This version check is lifted from the old Debian version. I doubt the old Debian version worked with Java 17 (see also #217).

The old RPM version had its own logic for Java version selection which IMHO tried to be way too clever about guessing the Java version. I have not retained the old RPM logic in the `systemd(1)` port. I don't think we should be in the business of trying to guess the best Java version. Rather I just use what's on the `PATH` and give the user the option to override it if necessary. Simple and straightforward.

Perhaps the version check from the old Debian version is overkill as well. I don't object to removing it. If it can be shown that the old RPM version worked with Java 17 and that it now doesn't work with the transition to `systemd(1)`, that is technically a regression and I will do a minimal fix in this PR. Otherwise I think Java 17 support for all platforms is best left to a future PR that covers the Java 17 issue comprehensively and systematically with testing on all platforms, `--add-opens`, etc.

Reply:
No regression as far as I can tell.  You're correct that the Debian code absolutely does not work with Java 17.  I was wrong in my thinking that the previous code allowed non-matching Java versions.

I had to modify the `/etc/init.d/jenkins` file to allow Java 17 with Jenkins 2.331, then had to modify `/etc/default/jenkins` to place JDK 17 in the `PATH` and set `JAVA_HOME` for JDK-17 and add `--enable-future-java`  to the Jenkins arguments.  That was enough to allow the controller to start, but that's already modifying files that are not intended to be modified.

I don't think we need to change the version check in this pull request.  When further work is done on Java 17 support, it will need changes in several locations.

Sorry for the distraction of a comment without the investigation that could have proven the comment was not needed.